### PR TITLE
Added Mingw-crt Support without breaking VS SDK

### PR DESCRIPTION
**Pull request by martell**
Title Says it all.

### CODE REVIEW

**Comment by martell** [comment]
Does anyone know how to remove the
Merge branch 'master' into mingw above?

I pulled from cocos2d/cocos2d-x to martell/cocos2d-x to update my master branch to the latest.
worked fine but I want to pull that into my mingw branch because it used the old master as the base?

The only way I could think of is merging my new pulled master into mingw.

Is there a better way to pull cocos2d/cocos2d-x master branch into martell/cocos2d-x mingw branch?

Hope I explained that well enough, I'm only getting to grips with the more detailed git usage

**Comment by dumganhar** [comment]
> Is there a better way to pull cocos2d/cocos2d-x master branch into martell/cocos2d/cocos2d-x mingw branch?

If your current branch is on `mingw`, you could use `git pull https://github.com/cocos2d/cocos2d-x.git master`.

**Comment by martell** [comment]
many thanks dumganhar. :)
That really helped :+1: 
That one was bugging me all day.

I essentially have a cocos remote hooked up like so.
git remote add cocos https://github.com/cocos2d/cocos2d-x.git
So could I just git pull cocos master? I thought I could only pull master into master?

**Comment by dumganhar** [comment]
> I essentially have a cocos remote hooked up like so.
> git remote add cocos https://github.com/cocos2d/cocos2d-x.git
> So could I just git pull cocos master? I thought I could only pull master into master?

Sure, actually, i do it in this way.

**Comment by martell** [comment]
Hi dumganhar it worked just like you said but the only problem i have is i still have a merge commit log after pulling?

Merge branch 'master' of https://github.com/cocos2d/cocos2d-x into mingw c8a9eff

I just rebased to remove the pull so this pull request should be fine now.
Still that is a strange one?

**Comment by dumganhar** [comment]
Yes, `rebase` could get rid of `Merge branch ...`, but sometimes `rebase` will make lots of conflicts.

**Comment by martell** [comment]
Ahh I see.
I seem to be a bit far from being a git expert.
How would you manage your branches dumganhar?

**Comment by dumganhar** [comment]
I just use some normal functionality of git, like `git checkout`,`git pull`, `git push`, `git merge`,`git log`,`git status`.
It will be easier if you use a UI tools of git. On Mac, you could use `Tower`(fee), `SourceTree`(free).
On Windows, `tortoisegit` is best.

**Comment by martell** [comment]
Could someone see this merged?

I'll do up a sample for using mingw32-make when I get the chance but as it doesn't break Visual studio versions I dont see a problem with merging it?

I'm using cmake which is why I haven't made a makefile sample yet

**Comment by dumganhar** [comment]
Hi @martell 

> I'm using cmake which is why I haven't made a makefile sample yet

Could you tell us how to use mingw32-make to build cocos2d-x?
I am not familiar with that. :(

**Comment by martell** [comment]
Hi dumganhar,

Don't worry we will sort this out :)

mingw32-make is essentially the same a running make on a linux box.
It will look for a makefile and run the script.
You can probably use the makefile from cocos2dx/proj.linux/Makefile
Just adjust it to use the platform/win32 instead of platform/linux
and set the CXXFlags at the top of the makefile to include all the definitions from the visual studio solution
 i.e. -DUNICODE -DCOCOS2DXWIN32_EXPORTS etc.

You will have to build libpng libjpeg libtiff and libwebp with mingw also.
I have several git repos which do this for you. ;)
Or can provide you with the libs prebuilt?

Which mingw toolchain version are you using?

**Comment by dumganhar** [comment]
In fact, i haven't install `mingw` toolchain yet.

**Comment by dumganhar** [comment]
I am just using cygwin for Android development.

**Comment by martell** [comment]
O right I see.
Well the version I use is here
http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/rubenvb/gcc-4.7-release/

i686-w64-mingw32-gcc-4.7.2-release-win32_rubenvb.7z

ruben has his build scripts on github.
They are the most stable of all mingw versions I have ever used.
He even has a build using llvm and clang if someone doesnt like gcc

You can crossbuild from linux to windows also if you want.
There is a cygwin build but its pointless to be honest because mingw defeats the purpose of having cygwin

**Comment by martell** [comment]
Any luck with your makefile?

**Comment by martell** [comment]
I figured out how to do add single commits to remove the merge commit msgs.
git cherry-pick 6992e97 :D ftw

File: cocos2dx/platform/win32/CCPlatformDefine.h [hunk 1]
```diff
@@ -1,6 +1,9 @@
 #ifndef __CCPLATFORMDEFINE_H__
 #define __CCPLATFORMDEFINE_H__
 
+#ifdef __MINGW32__
+#include <string.h>
+#endif
```
Comment (dumganhar) [inline cocos2dx/platform/win32/CCPlatformDefine.h:6]: Why to add `#include <string.h>` here?
Could it be added to `win32/CCStdC.h`?

**Comment by martell** [comment]
No I don't think it can.
When I compiled that way it failed

I just copied the structure you guys already have
linux/CCPlatformDefine.h  ?

Another thing is that in CCPlatform define you guys have a #ifdef for _WINDOWS that doesn't exist on mingw
I added it to my makefile instead of changing your code. 
The WIN32 define is fine. 
Other than that it should all work fine

**Comment by minggo** [comment]
I think add mingw support will make things too complex.
Windows is not a target platform for releasing, why make it complex?

**Comment by martell** [comment]
It doesn't really affect much. Just two header files.
#ifdef MINGW in a few places is all it does.

I would support mingw because there are now builds of clang and llvm for the mingw runtimes.
hint. emscripten ;)

To be honest I believe they are more stable that the MS compilers(from personal experience)
They map across to android and all other platforms that use gcc without and MS specific issues

This way we don't have to install visual studio to do development on windows if we don't want to..
You can use another ide such as sumblime text

But all this is down to the programmers preference.

**Pull request merged**